include/circt/Conversion/Passes.td

```diff
@@ -200,7 +200,14 @@ def LoopScheduleToCalyx : Pass<"lower-loopschedule-to-calyx", "mlir::ModuleOp">
 // PipelineTHW
 //===----------------------------------------------------------------------===//
 
-def PipelineToHW : Pass<"lower-pipeline-to-hw", "hw::HWModuleOp"> {
+// TODO: @mortbopet: There is a possible non-neglible speedup that can be achieved
+// here by allowing this pass to run on a per-hwmodule/whatever container the
+// pipeline is nested within-granularity. However, this conversion adds (and removes)
+// new modules to the top-level mlir::ModuleOp scope, which technically violates
+// hw::HWModuleLike's IsolatedFromAbove (and thus has previously caused
```

Comment:
FWIW don't think whether it's `IsolatedFromAbove` is relevant, it's that passes cannot change the containing operation (or even the current one in ways that's visible/possibly-breaking to other operations, such as adding/removing ports of a `hw.module` from a pass running on `hw.module`'s).

Reply:
> passes cannot change the containing operation

I see `IsolatedFromAbove` as a trait inherently linked to the passmanager, and how passes are scheduled. IsolatedFromAbove allows multi-threaded execution, which in turn is what "allows" these errors to happen.